### PR TITLE
CupertinoTabBar

### PR DESCRIPTION
**Pull request by xster**
#9926

Also moved some cupertino colors into a common file

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  /// The size of all of the [BottomNavigationBarItem] icons.
+  ///
+  /// This value is used to to configure the [IconTheme] for the navigation
+  /// bar. When a [BottomNavigationBarItem.icon] widget is not an [Icon] the widget
+  /// should configure itself to match the icon theme's size and color.
+  final double iconSize;
+
+  @override
+  Widget build(BuildContext context) {
+    final bool addBlur = backgroundColor.opacity < 1.0;
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: it's a trivial issue but `alpha != 0xFF` would probably be a couple of instructions faster

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    Widget returnWidget;
+
+    returnWidget = new DecoratedBox(
+      decoration: new BoxDecoration(
+        border: const Border(
+          top: const BorderSide(
+            color: const Color(0x4C000000),
+            width: 0.5, // One physical pixel.
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: 0.0 will give you a hairline that's always exactly 1 physical pixel.

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  /// should configure itself to match the icon theme's size and color.
+  final double iconSize;
+
+  @override
+  Widget build(BuildContext context) {
+    final bool addBlur = backgroundColor.opacity < 1.0;
+
+    Widget returnWidget;
+
+    returnWidget = new DecoratedBox(
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: prefer `result` for the return value variable.

prefer `Widget result = ...` so that the variable is never null, it'll make it easier for us to switch to non-nullable types later

File: packages/flutter/lib/src/cupertino/button.dart [hunk 4]
```diff
@@ -25,7 +25,7 @@ final TextStyle _kDisabledButtonTextStyle = _kButtonTextStyle.copyWith(
 );
 
 final TextStyle _kBackgroundButtonTextStyle = _kButtonTextStyle.copyWith(
-  color: _kWhite,
+  color: CupertinoColors.activeBlue,
```
Comment (abarth) [inline packages/flutter/lib/src/cupertino/button.dart]: White became blue here.

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      );
+    }
+
+    return returnWidget;
+  }
+
+  List<Widget> _buildTabItems() {
+    final List<Widget> returnWidgets = <Widget>[];
+
+    for (int i=0; i<items.length; i++) {
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: spaces around operators
prefer `i += 1` (my preference) or `++i` (adam's preference) to `i++`

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+      returnWidgets.add(
+        _wrapActiveItem(
+
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: unexpected blank line

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 5 above)
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: prefer `index` to `i` (avoid abbreviations where possible)

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                mainAxisAlignment: MainAxisAlignment.end,
+                children: <Widget> [
+                  new Expanded(child: new Center(child: items[i].icon)),
+                  items[i].title,
+                  const Padding(padding: const EdgeInsets.only(bottom: 4.0)),
+                ],
+              ),
+            ),
+          ),
+          active: i == currentIndex
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: nit: trailing comma

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 1 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: Done

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 2 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: waaaaaaat, awesome

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+              onTap: () {
+                if (onTap != null)
+                  onTap(i);
+              },
+              child: new Column(
+                mainAxisAlignment: MainAxisAlignment.end,
+                children: <Widget> [
+                  new Expanded(child: new Center(child: items[i].icon)),
+                  items[i].title,
+                  const Padding(padding: const EdgeInsets.only(bottom: 4.0)),
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: if you just want a 4.0-high box, prefer `SizedBox` to `Padding`.

File: packages/flutter/lib/src/cupertino/button.dart (hunk 4 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/button.dart]: ah shit, my bad

File: packages/flutter/lib/src/cupertino/button.dart [hunk 9]
```diff
@@ -5,9 +5,9 @@
 import 'package:flutter/foundation.dart';
 import 'package:flutter/widgets.dart';
 
+import 'colors.dart';
+
 // TODO(xster): move this to a common Cupertino color palette with the next yak.
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/button.dart]: remove the TODO?

File: packages/flutter/lib/src/cupertino/colors.dart [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  /// Opaque black color. Used for texts against light backgrounds.
+  static const Color black = const Color(0xFF000000);
+
+  /// Used in iOS 10 for light background fills such as the chat bubble background.
+  static const Color lightBackgroundGray = const Color(0xFFE5E5EA);
+
+  /// Used in iOS 10 for unselected selectables such as tab bar items in their
+  /// inactive state.
+  ///
+  /// Note it's not the same gray as disabled buttons etc.
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/colors.dart]: avoid the word "note" -- see style guide https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

File: packages/flutter/lib/src/cupertino/colors.dart [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  static const Color inactiveGray = const Color(0xFF929292);
+
+  /// Used for iOS 10 for destructive actions such as the delete actions in
+  /// table view cells and dialogs.
+  ///
+  /// Note it's not the same red as the camera shutter or springboard icon notifications
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/colors.dart]: ditto

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 3 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: done

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 6 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: oops sorry

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 5 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: done

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 7 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: done

File: packages/flutter/lib/src/cupertino/button.dart (hunk 9 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/button.dart]: done

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+import 'dart:ui';
+import 'package:flutter/foundation.dart';
+
+import 'framework.dart';
+
+/// An interactive button within either material's [BottomNavigationBar]
+/// or the iOS themed [CupertinoTabBar] with an icon and title.
+///
+/// Should not be used on its own but embedded in one of the bottom navigation
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: rather than this paragraph I would say something like "This class is rarely used in isolation".

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  ///
+  /// If the navigation bar's type is [BottomNavigationBarType.shifting], then
+  /// the entire bar is flooded with the [backgroundColor] when this item is
+  /// tapped.
+  ///
+  /// Not used for [CupertinoTabBar]. Control the invariant bar color directly
+  /// via [CupertinoTabBar.backgroundColor].
+  ///
+  /// See also:
+  ///  * [Icon.backgroundColor] and [ImageIcon.backgroundColor] to control the foreground color of
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: dartdocs' markdown needs a blank line between "See also" and the bullet point.

Comment (Hixie) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: I'm hoping this comment is wrong, because otherwise we really picked some bad names for those fields...

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+/// Should not be used on its own but embedded in one of the bottom navigation
+/// widgets above.
+///
+/// See also:
+///
+///  * [BottomNavigationBar]
+///  * <https://material.google.com/components/bottom-navigation.html>
+///  * [CupertinoTabBar]
+///  * <https://developer.apple.com/ios/human-interface-guidelines/ui-bars/tab-bars>
+class BottomNavigationBarItem {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: can we make this class const while we're at it?

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+  /// Creates an item that is used with [BottomNavigationBar.items].
+  ///
+  /// The arguments [icon] and [title] should not be null.
+  BottomNavigationBarItem({
+    @required this.icon,
+    @required this.title,
+    this.backgroundColor
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: nit: trailing comma

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart [hunk 16]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+import 'package:flutter/foundation.dart';
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: blank line between dart: imports and package: imports
dart:ui should always be imported `as ui` or, for the most common classes, should explicitly say `show Color` et al.

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+    final List<Widget> returnWidgets = <Widget>[];
+
+    for (int i=0; i<items.length; i++) {
+      returnWidgets.add(
+        _wrapActiveItem(
+
+          new Expanded(
+            child: new GestureDetector(
+              onTap: () {
+                if (onTap != null)
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: Do we want to detect taps even when there's no onTap handler?

Or should we instead say `onTap: onTap != null ? () { onTap(i) } : null,` ?

**Comment by Hixie** [comment]
Tests?

**Comment by eseidelGoogle** [comment]
FYI @FaisalAbid you may be interested in this.

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 8 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: ah ok, done (in the coming commit). What's the nuance though? Should we add some more implications doc to the Padding class in the 'design discussion' section to not add empty paddings?

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 17 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: good catch. Thanks

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart (hunk 16 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: uh oh, I did it in a bunch of places. Corrected

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart (hunk 15 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: Done

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart (hunk 14 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: good point, done

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart (hunk 13 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: doh, a vscode refactor drive-by
thanks

Comment (xster) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: ah ok, done

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart (hunk 12 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: done

File: packages/flutter/lib/src/cupertino/colors.dart (hunk 11 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/colors.dart]: done

File: packages/flutter/lib/src/cupertino/colors.dart (hunk 10 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/colors.dart]: done

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 5 above)
Comment (xster) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: done

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 8 above)
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: It's mostly just cleaner, and infinitesimally faster. But yeah, maybe we should mention it in passing in the docs in a "best practices" section for the Padding API docs. Feel free to add that in this PR.

**Comment by Hixie** [comment]
LGTM
